<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<title>Exploring NMF and other views of high-dimensional multivariate data • multivariateEpigenomics</title>
<script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet">
<script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/font-awesome-6.5.2/css/all.min.css" rel="stylesheet">
<link href="../deps/font-awesome-6.5.2/css/v4-shims.min.css" rel="stylesheet">
<script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Exploring NMF and other views of high-dimensional multivariate data">
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-light" data-bs-theme="light" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">multivariateEpigenomics</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">0.0.1</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="nav-item"><a class="nav-link" href="../reference/index.html">Reference</a></li>
<li class="active nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-articles" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">Articles</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-articles">
<li><a class="dropdown-item" href="../articles/nmflook.html">Exploring NMF and other views of high-dimensional multivariate data</a></li>
  </ul>
</li>
      </ul>
<ul class="navbar-nav">
<li class="nav-item"><form class="form-inline" role="search">
 <input class="form-control" type="search" name="search-input" id="search-input" autocomplete="off" aria-label="Search site" placeholder="Search for" data-search-index="../search.json">
</form></li>
<li class="nav-item"><a class="nav-link" href="https://github.com/vjcitn/multivariateEpigenomics/" aria-label="GitHub"><span class="fa fab fa-github fa-lg"></span></a></li>
      </ul>
</div>


  </div>
</nav><div class="container template-article">




<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">

      <h1>Exploring NMF and other views of high-dimensional multivariate data</h1>
                        <h4 data-toc-skip class="author">Vincent J. Carey, stvjc at channing.harvard.edu</h4>
            
            <h4 data-toc-skip class="date">January 25, 2025</h4>
      
      <small class="dont-index">Source: <a href="https://github.com/vjcitn/multivariateEpigenomics/blob/HEAD/vignettes/nmflook.Rmd"><code>vignettes/nmflook.Rmd</code></a></small>
      <div class="d-none name"><code>nmflook.Rmd</code></div>
    </div>

    
    
<div class="section level2">
<h2 id="overview">Overview<a class="anchor" aria-label="anchor" href="#overview"></a>
</h2>
<p>To improve my understanding of the role of the HALS algorithm for NMF analysis of genomic data, I work with the ALL microarray dataset.</p>
<pre><code>     Sabina Chiaretti, Xiaochun Li, Robert Gentleman, Antonella Vitale,
     Marco Vignetti, Franco Mandelli, Jerome Ritz, and Robin Foa Gene
     expression profile of adult T-cell acute lymphocytic leukemia
     identifies distinct subsets of patients with different response to
     therapy and survival. Blood, 1 April 2004, Vol. 103, No. 7.</code></pre>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># to avoid conflicting symbol warnings</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/warning.html" class="external-link">suppressWarnings</a></span><span class="op">(</span><span class="op">{</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">ALL</span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/HenrikBengtsson/matrixStats" class="external-link">matrixStats</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">hgu95av2.db</span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://bioconductor.org/packages/SummarizedExperiment" class="external-link">SummarizedExperiment</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://www.stat.berkeley.edu/~breiman/RandomForests/" class="external-link">randomForest</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/friendly/ggbiplot" class="external-link">ggbiplot</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/vjcitn/multivariateEpigenomics">multivariateEpigenomics</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="http://renozao.github.io/NMF/" class="external-link">NMF</a></span><span class="op">)</span></span>
<span><span class="op">}</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="data-acquisition-and-setup">Data acquisition and setup<a class="anchor" aria-label="anchor" href="#data-acquisition-and-setup"></a>
</h2>
<div class="section level3">
<h3 id="filtering-and-feature-annotation">Filtering and feature annotation<a class="anchor" aria-label="anchor" href="#filtering-and-feature-annotation"></a>
</h3>
<p>I reduce the dataset down to two patient types: those with BCR/ABL fusion and those negative for any known cytogenetic feature used in prognosis at the time of the study.</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">ALL</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="va">ALL</span><span class="op">)</span></span>
<span><span class="va">fuvn</span> <span class="op">=</span> <span class="va">ALL</span><span class="op">[</span>, <span class="fu"><a href="https://rdrr.io/r/base/which.html" class="external-link">which</a></span><span class="op">(</span><span class="va">ALL</span><span class="op">$</span><span class="va">mol.biol</span> <span class="op"><a href="https://rdrr.io/r/base/match.html" class="external-link">%in%</a></span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"BCR/ABL"</span>, <span class="st">"NEG"</span><span class="op">)</span><span class="op">)</span><span class="op">]</span></span>
<span><span class="va">fuvn</span><span class="op">$</span><span class="va">mol.biol</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/factor.html" class="external-link">factor</a></span><span class="op">(</span><span class="va">fuvn</span><span class="op">$</span><span class="va">mol.biol</span><span class="op">)</span></span></code></pre></div>
<p>I filter the features down to those in the upper quantile of between-patient variation.</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/HenrikBengtsson/matrixStats" class="external-link">matrixStats</a></span><span class="op">)</span></span>
<span><span class="va">ss</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/matrixStats/man/rowSds.html" class="external-link">rowSds</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/Biobase/man/exprs.html" class="external-link">exprs</a></span><span class="op">(</span><span class="va">fuvn</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="co">#&gt; sum(ss&gt;.5)</span></span>
<span><span class="co">#[1] 3182</span></span>
<span><span class="va">ffuvn</span> <span class="op">=</span> <span class="va">fuvn</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/which.html" class="external-link">which</a></span><span class="op">(</span><span class="va">ss</span><span class="op">&gt;</span><span class="fl">.5</span><span class="op">)</span>,<span class="op">]</span></span></code></pre></div>
<p>Transform the data representation to SummarizedExperiment and rename the features with HUGO symbols.</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">hgu95av2.db</span><span class="op">)</span></span>
<span><span class="va">mapas</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/AnnotationDbi/man/AnnotationDb-class.html" class="external-link">mapIds</a></span><span class="op">(</span><span class="va">hgu95av2.db</span>, keys<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">ffuvn</span><span class="op">)</span>, column<span class="op">=</span><span class="st">"SYMBOL"</span>, keytype<span class="op">=</span><span class="st">"PROBEID"</span><span class="op">)</span></span>
<span><span class="co"># as = affy to symbol</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://bioconductor.org/packages/SummarizedExperiment" class="external-link">SummarizedExperiment</a></span><span class="op">)</span></span>
<span><span class="va">ffuvnse</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/methods/as.html" class="external-link">as</a></span><span class="op">(</span><span class="va">ffuvn</span>, <span class="st">"SummarizedExperiment"</span><span class="op">)</span></span>
<span><span class="va">ffuvnse</span></span></code></pre></div>
<pre><code><span><span class="co">## class: SummarizedExperiment </span></span>
<span><span class="co">## dim: 3182 111 </span></span>
<span><span class="co">## metadata(3): experimentData annotation protocolData</span></span>
<span><span class="co">## assays(1): exprs</span></span>
<span><span class="co">## rownames(3182): 1005_at 1007_s_at ... AFFX-YEL021w/URA3_at</span></span>
<span><span class="co">##   AFFX-YEL024w/RIP1_at</span></span>
<span><span class="co">## rowData names(0):</span></span>
<span><span class="co">## colnames(111): 01005 01010 ... 83001 LAL4</span></span>
<span><span class="co">## colData names(21): cod diagnosis ... f.u date.last.seen</span></span></code></pre>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">bad</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/which.html" class="external-link">which</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/NA.html" class="external-link">is.na</a></span><span class="op">(</span><span class="va">mapas</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">pbad</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">mapas</span><span class="op">)</span><span class="op">[</span><span class="va">bad</span><span class="op">]</span></span>
<span><span class="va">exclprobes</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/sets.html" class="external-link">intersect</a></span><span class="op">(</span><span class="va">pbad</span>, <span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">ffuvnse</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">badinds</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/which.html" class="external-link">which</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">ffuvnse</span><span class="op">)</span> <span class="op"><a href="https://rdrr.io/r/base/match.html" class="external-link">%in%</a></span> <span class="va">exclprobes</span><span class="op">)</span></span>
<span><span class="va">ffuvnse</span> <span class="op">=</span> <span class="va">ffuvnse</span><span class="op">[</span><span class="op">-</span><span class="va">badinds</span>,<span class="op">]</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">rowData</a></span><span class="op">(</span><span class="va">ffuvnse</span><span class="op">)</span> <span class="op">=</span> <span class="fu">DataFrame</span><span class="op">(</span>probeid<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">ffuvnse</span><span class="op">)</span>, sym<span class="op">=</span><span class="va">mapas</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">ffuvnse</span><span class="op">)</span><span class="op">]</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">ffuvnse</span><span class="op">)</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">rowData</a></span><span class="op">(</span><span class="va">ffuvnse</span><span class="op">)</span><span class="op">$</span><span class="va">sym</span></span>
<span><span class="va">ffuvnse</span></span></code></pre></div>
<pre><code><span><span class="co">## class: SummarizedExperiment </span></span>
<span><span class="co">## dim: 3131 111 </span></span>
<span><span class="co">## metadata(3): experimentData annotation protocolData</span></span>
<span><span class="co">## assays(1): exprs</span></span>
<span><span class="co">## rownames(3131): DUSP1 DDR1 ... STAT1 TFRC</span></span>
<span><span class="co">## rowData names(2): probeid sym</span></span>
<span><span class="co">## colnames(111): 01005 01010 ... 83001 LAL4</span></span>
<span><span class="co">## colData names(21): cod diagnosis ... f.u date.last.seen</span></span></code></pre>
</div>
</div>
<div class="section level2">
<h2 id="isolation-of-informative-features-using-random-forests">Isolation of informative features using random forests<a class="anchor" aria-label="anchor" href="#isolation-of-informative-features-using-random-forests"></a>
</h2>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://www.stat.berkeley.edu/~breiman/RandomForests/" class="external-link">randomForest</a></span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/Random.html" class="external-link">set.seed</a></span><span class="op">(</span><span class="fl">1234</span><span class="op">)</span> <span class="co"># make reproducible</span></span>
<span><span class="va">ldf</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/data.frame.html" class="external-link">data.frame</a></span><span class="op">(</span>molb<span class="op">=</span><span class="va">ffuvnse</span><span class="op">$</span><span class="va">mol.biol</span>, <span class="fu"><a href="https://rdrr.io/r/base/t.html" class="external-link">t</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">assay</a></span><span class="op">(</span><span class="va">ffuvnse</span><span class="op">)</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">rf1</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/randomForest/man/randomForest.html" class="external-link">randomForest</a></span><span class="op">(</span><span class="va">molb</span><span class="op">~</span><span class="va">.</span>, data<span class="op">=</span><span class="va">ldf</span>, importance<span class="op">=</span><span class="cn">TRUE</span><span class="op">)</span></span>
<span><span class="va">rf1</span></span></code></pre></div>
<pre><code><span><span class="co">## </span></span>
<span><span class="co">## Call:</span></span>
<span><span class="co">##  randomForest(formula = molb ~ ., data = ldf, importance = TRUE) </span></span>
<span><span class="co">##                Type of random forest: classification</span></span>
<span><span class="co">##                      Number of trees: 500</span></span>
<span><span class="co">## No. of variables tried at each split: 55</span></span>
<span><span class="co">## </span></span>
<span><span class="co">##         OOB estimate of  error rate: 11.71%</span></span>
<span><span class="co">## Confusion matrix:</span></span>
<span><span class="co">##         BCR/ABL NEG class.error</span></span>
<span><span class="co">## BCR/ABL      28   9  0.24324324</span></span>
<span><span class="co">## NEG           4  70  0.05405405</span></span></code></pre>
<p>Even though the overall OOB misclassification estimate is attractive, the confusion matrix is not very appealing.</p>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/randomForest/man/varImpPlot.html" class="external-link">varImpPlot</a></span><span class="op">(</span><span class="va">rf1</span><span class="op">)</span></span></code></pre></div>
<p><img src="nmflook_files/figure-html/lkvari-1.png" width="700"></p>
<p>Let’s take the top 15 features for multivariate analysis.</p>
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">im</span> <span class="op">=</span> <span class="va">rf1</span><span class="op">$</span><span class="va">importance</span></span>
<span><span class="va">top</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">im</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/order.html" class="external-link">order</a></span><span class="op">(</span><span class="va">im</span><span class="op">[</span>,<span class="fl">3</span><span class="op">]</span>, decreasing<span class="op">=</span><span class="cn">TRUE</span><span class="op">)</span>,<span class="op">]</span><span class="op">)</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">15</span><span class="op">]</span></span>
<span><span class="va">top</span></span></code></pre></div>
<pre><code><span><span class="co">##  [1] "ABL1.1"   "ABL1.3"   "ALDH1A1"  "KLF9"     "EPS8"     "ABL1"    </span></span>
<span><span class="co">##  [7] "PON2"     "MARCKS"   "SEMA6A"   "TUBA4A.1" "MTSS1"    "CNN3"    </span></span>
<span><span class="co">## [13] "CDK14"    "GAB1.2"   "ECM1"</span></span></code></pre>
</div>
<div class="section level2">
<h2 id="multivariate-exploration-1-pca">Multivariate exploration 1: PCA<a class="anchor" aria-label="anchor" href="#multivariate-exploration-1-pca"></a>
</h2>
<p>Basic computation:</p>
<div class="sourceCode" id="cb14"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">pp</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/stats/prcomp.html" class="external-link">prcomp</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/data.matrix.html" class="external-link">data.matrix</a></span><span class="op">(</span><span class="va">ldf</span><span class="op">[</span>,<span class="va">top</span><span class="op">]</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p>Overview:</p>
<div class="sourceCode" id="cb15"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/graphics/pairs.html" class="external-link">pairs</a></span><span class="op">(</span><span class="va">pp</span><span class="op">$</span><span class="va">x</span><span class="op">[</span>,<span class="fl">1</span><span class="op">:</span><span class="fl">4</span><span class="op">]</span>, col<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/factor.html" class="external-link">factor</a></span><span class="op">(</span><span class="va">ldf</span><span class="op">$</span><span class="va">molb</span><span class="op">)</span>, pch<span class="op">=</span><span class="fl">19</span>, cex<span class="op">=</span><span class="fl">.4</span><span class="op">)</span></span></code></pre></div>
<p><img src="nmflook_files/figure-html/lkpcs-1.png" width="700"></p>
<p>A simple biplot to show some details of the projection (first two components)</p>
<div class="sourceCode" id="cb16"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="http://friendly.github.io/ggbiplot/reference/ggbiplot.html" class="external-link">ggbiplot</a></span><span class="op">(</span><span class="va">pp</span><span class="op">)</span> <span class="op">+</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_point.html" class="external-link">geom_point</a></span><span class="op">(</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span>colour<span class="op">=</span><span class="va">ldf</span><span class="op">$</span><span class="va">molb</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p><img src="nmflook_files/figure-html/dopc1b-1.png" width="700"> Try 2 and 3:</p>
<div class="sourceCode" id="cb17"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="http://friendly.github.io/ggbiplot/reference/ggbiplot.html" class="external-link">ggbiplot</a></span><span class="op">(</span><span class="va">pp</span>, <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">2</span>,<span class="fl">3</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_point.html" class="external-link">geom_point</a></span><span class="op">(</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span>colour<span class="op">=</span><span class="va">ldf</span><span class="op">$</span><span class="va">molb</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p><img src="nmflook_files/figure-html/dopc2-1.png" width="700"></p>
<p>While PC1 seems to provide information to discriminate BCR/ABL fusion from other ALL negative for prognostic cytogenetic features, PC2 and PC3 may be helping to isolate genes that discriminate different transcriptional patterns (and possibly different clinical sequelae) within BCR/ABL fusion.</p>
<p>We’ll be comparing this data reduction method to NMF. NMF results are frequently presented as heatmaps. Here’s one approach for PCA:</p>
<div class="sourceCode" id="cb18"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/NMF/man/aheatmap.html" class="external-link">aheatmap</a></span><span class="op">(</span><span class="va">pp</span><span class="op">$</span><span class="va">x</span><span class="op">[</span>,<span class="fl">1</span><span class="op">:</span><span class="fl">4</span><span class="op">]</span>, scale<span class="op">=</span><span class="st">"none"</span>, annRow<span class="op">=</span><span class="va">ldf</span><span class="op">$</span><span class="va">molb</span><span class="op">)</span></span></code></pre></div>
<p><img src="nmflook_files/figure-html/lkpcsheat-1.png" width="700"></p>
<div class="sourceCode" id="cb19"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/NMF/man/aheatmap.html" class="external-link">aheatmap</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/t.html" class="external-link">t</a></span><span class="op">(</span><span class="va">pp</span><span class="op">$</span><span class="va">rot</span><span class="op">[</span><span class="va">top</span>,<span class="fl">1</span><span class="op">:</span><span class="fl">4</span><span class="op">]</span><span class="op">)</span>, scale<span class="op">=</span><span class="st">"none"</span>, Rowv<span class="op">=</span><span class="cn">NA</span>, Colv<span class="op">=</span><span class="cn">NA</span><span class="op">)</span></span></code></pre></div>
<p><img src="nmflook_files/figure-html/lkpcrot-1.png" width="700"></p>
<p>At a minimum, this shows that low values of PC1 discriminate BCR/ABL fusion reasonably well.</p>
<p>To confirm this, one could work with samples from the TARGET study using Bioconductor’s <a href="https://bioconductor.org/packages/release/bioc/html/GenomicDataCommons.html" class="external-link">GenomicDataCommons</a> package.</p>
</div>
<div class="section level2">
<h2 id="multivariate-exploration-2-nmf">Multivariate exploration 2: NMF<a class="anchor" aria-label="anchor" href="#multivariate-exploration-2-nmf"></a>
</h2>
<p>How does non-negative matrix factorization (NMF) help us explore expression patterns in this data? The matrix <math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><msub><mi>X</mi><mrow><mi>n</mi><mo>×</mo><mi>p</mi></mrow></msub><annotation encoding="application/x-tex">X_{n \times p}</annotation></semantics></math> of expression values (rows correspond to patients, columns to genes) is factored as <math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><mi>X</mi><mo>≈</mo><mi>W</mi><msup><mi>H</mi><mi>t</mi></msup></mrow><annotation encoding="application/x-tex">X \approx WH^t</annotation></semantics></math> with the rank of <math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mi>W</mi><annotation encoding="application/x-tex">W</annotation></semantics></math> set at <math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><mi>r</mi><mo>&lt;</mo><mo>&lt;</mo><mi>p</mi></mrow><annotation encoding="application/x-tex">r &lt;&lt; p</annotation></semantics></math>.</p>
<div class="section level3">
<h3 id="gaujoux-package">Gaujoux’ package<a class="anchor" aria-label="anchor" href="#gaujoux-package"></a>
</h3>
<p>We’ll use a CRAN package for NMF by Renaud Gaujoux and set <math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mi>r</mi><annotation encoding="application/x-tex">r</annotation></semantics></math> to 4.</p>
<div class="sourceCode" id="cb20"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="http://renozao.github.io/NMF/" class="external-link">NMF</a></span><span class="op">)</span></span>
<span><span class="va">nmf1</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/NMF/man/nmf.html" class="external-link">nmf</a></span><span class="op">(</span><span class="va">ldf</span><span class="op">[</span>, <span class="va">top</span><span class="op">]</span>, <span class="fl">4</span><span class="op">)</span></span></code></pre></div>
<p>The W and H matrices can be visualized in heatmaps:</p>
<div class="sourceCode" id="cb21"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/NMF/man/heatmaps.html" class="external-link">basismap</a></span><span class="op">(</span><span class="va">nmf1</span>, annRow<span class="op">=</span><span class="va">ldf</span><span class="op">$</span><span class="va">molb</span><span class="op">)</span></span></code></pre></div>
<p><img src="nmflook_files/figure-html/lkw-1.png" width="700"></p>
<div class="sourceCode" id="cb22"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/NMF/man/heatmaps.html" class="external-link">coefmap</a></span><span class="op">(</span><span class="va">nmf1</span><span class="op">)</span></span></code></pre></div>
<p><img src="nmflook_files/figure-html/lkw-2.png" width="700"></p>
</div>
<div class="section level3">
<h3 id="gregg-implementation-of-hals-in-julia">Gregg implementation of HALS in Julia<a class="anchor" aria-label="anchor" href="#gregg-implementation-of-hals-in-julia"></a>
</h3>
<div class="sourceCode" id="cb23"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/vjcitn/multivariateEpigenomics">multivariateEpigenomics</a></span><span class="op">)</span></span>
<span><span class="va">nmf2</span> <span class="op">=</span> <span class="fu"><a href="../reference/nmf_HALS.html">nmf_HALS</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/data.matrix.html" class="external-link">data.matrix</a></span><span class="op">(</span><span class="va">ldf</span><span class="op">[</span>, <span class="va">top</span><span class="op">]</span><span class="op">)</span>, <span class="fl">4</span><span class="op">)</span></span></code></pre></div>
<p>Here’s an analog of basismap for the HALS algorithm:</p>
<div class="sourceCode" id="cb24"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/NMF/man/aheatmap.html" class="external-link">aheatmap</a></span><span class="op">(</span><span class="va">nmf2</span><span class="op">$</span><span class="va">W</span>, scale<span class="op">=</span><span class="st">"r1"</span>, annRow<span class="op">=</span><span class="va">ldf</span><span class="op">$</span><span class="va">molb</span><span class="op">)</span></span></code></pre></div>
<p><img src="nmflook_files/figure-html/dobashals-1.png" width="700"></p>
<p>Coefficients:</p>
<div class="sourceCode" id="cb25"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/NMF/man/aheatmap.html" class="external-link">aheatmap</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/t.html" class="external-link">t</a></span><span class="op">(</span><span class="va">nmf2</span><span class="op">$</span><span class="va">H</span><span class="op">)</span>, scale<span class="op">=</span><span class="st">"c1"</span>, Rowv<span class="op">=</span><span class="cn">NA</span>, Colv<span class="op">=</span><span class="cn">TRUE</span><span class="op">)</span></span></code></pre></div>
<p><img src="nmflook_files/figure-html/docohals-1.png" width="700"></p>
</div>
</div>
<div class="section level2">
<h2 id="comparing-discriminative-capacities-of-the-three-approaches">Comparing discriminative capacities of the three approaches<a class="anchor" aria-label="anchor" href="#comparing-discriminative-capacities-of-the-three-approaches"></a>
</h2>
<p>We’ll use random forests with the data reductions.</p>
<div class="section level3">
<h3 id="four-pcs">Four PCs<a class="anchor" aria-label="anchor" href="#four-pcs"></a>
</h3>
<div class="sourceCode" id="cb26"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">newdf1</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/data.frame.html" class="external-link">data.frame</a></span><span class="op">(</span>molb<span class="op">=</span><span class="va">ldf</span><span class="op">$</span><span class="va">molb</span>, <span class="va">pp</span><span class="op">$</span><span class="va">x</span><span class="op">[</span>,<span class="fl">1</span><span class="op">:</span><span class="fl">4</span><span class="op">]</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">newdf1</span>,<span class="fl">3</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">##          molb       PC1         PC2        PC3        PC4</span></span>
<span><span class="co">## 01005 BCR/ABL -5.286390  0.28812041 -0.5818160 -0.4489509</span></span>
<span><span class="co">## 01010     NEG  1.283034 -0.74729496  0.4412039  1.2070504</span></span>
<span><span class="co">## 03002 BCR/ABL -6.268100 -0.09124536 -0.5753392  0.3960636</span></span></code></pre>
<div class="sourceCode" id="cb28"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">nr1</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/randomForest/man/randomForest.html" class="external-link">randomForest</a></span><span class="op">(</span><span class="va">molb</span><span class="op">~</span><span class="va">.</span>, data<span class="op">=</span><span class="va">newdf1</span><span class="op">)</span></span>
<span><span class="va">nr1</span></span></code></pre></div>
<pre><code><span><span class="co">## </span></span>
<span><span class="co">## Call:</span></span>
<span><span class="co">##  randomForest(formula = molb ~ ., data = newdf1) </span></span>
<span><span class="co">##                Type of random forest: classification</span></span>
<span><span class="co">##                      Number of trees: 500</span></span>
<span><span class="co">## No. of variables tried at each split: 2</span></span>
<span><span class="co">## </span></span>
<span><span class="co">##         OOB estimate of  error rate: 7.21%</span></span>
<span><span class="co">## Confusion matrix:</span></span>
<span><span class="co">##         BCR/ABL NEG class.error</span></span>
<span><span class="co">## BCR/ABL      33   4  0.10810811</span></span>
<span><span class="co">## NEG           4  70  0.05405405</span></span></code></pre>
</div>
<div class="section level3">
<h3 id="four-basis-elements-from-gajoux-nmf">Four basis elements from Gajoux NMF<a class="anchor" aria-label="anchor" href="#four-basis-elements-from-gajoux-nmf"></a>
</h3>
<div class="sourceCode" id="cb30"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">newdf2</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/data.frame.html" class="external-link">data.frame</a></span><span class="op">(</span>molb<span class="op">=</span><span class="va">ldf</span><span class="op">$</span><span class="va">molb</span>, <span class="fu"><a href="https://rdrr.io/pkg/NMF/man/basis-coef-methods.html" class="external-link">basis</a></span><span class="op">(</span><span class="va">nmf1</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">newdf2</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">##          molb       X1        X2       X3       X4</span></span>
<span><span class="co">## 01005 BCR/ABL 7.353671  9.664595 4.435785 4.127642</span></span>
<span><span class="co">## 01010     NEG 3.339959  5.730598 5.787085 5.411212</span></span>
<span><span class="co">## 03002 BCR/ABL 8.169915 11.571760 2.302937 3.888486</span></span>
<span><span class="co">## 04007     NEG 4.803022 10.131980 3.362568 4.483045</span></span>
<span><span class="co">## 04008     NEG 3.487283  6.252317 5.691256 3.190453</span></span>
<span><span class="co">## 04010     NEG 2.931594  6.377346 6.169460 4.316055</span></span></code></pre>
<div class="sourceCode" id="cb32"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">nr2</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/randomForest/man/randomForest.html" class="external-link">randomForest</a></span><span class="op">(</span><span class="va">molb</span><span class="op">~</span><span class="va">.</span>, data<span class="op">=</span><span class="va">newdf2</span><span class="op">)</span></span>
<span><span class="va">nr2</span></span></code></pre></div>
<pre><code><span><span class="co">## </span></span>
<span><span class="co">## Call:</span></span>
<span><span class="co">##  randomForest(formula = molb ~ ., data = newdf2) </span></span>
<span><span class="co">##                Type of random forest: classification</span></span>
<span><span class="co">##                      Number of trees: 500</span></span>
<span><span class="co">## No. of variables tried at each split: 2</span></span>
<span><span class="co">## </span></span>
<span><span class="co">##         OOB estimate of  error rate: 19.82%</span></span>
<span><span class="co">## Confusion matrix:</span></span>
<span><span class="co">##         BCR/ABL NEG class.error</span></span>
<span><span class="co">## BCR/ABL      25  12   0.3243243</span></span>
<span><span class="co">## NEG          10  64   0.1351351</span></span></code></pre>
</div>
<div class="section level3">
<h3 id="four-basis-elements-from-gregg-nmf">Four basis elements from Gregg NMF<a class="anchor" aria-label="anchor" href="#four-basis-elements-from-gregg-nmf"></a>
</h3>
<div class="sourceCode" id="cb34"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">newdf3</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/data.frame.html" class="external-link">data.frame</a></span><span class="op">(</span>molb<span class="op">=</span><span class="va">ldf</span><span class="op">$</span><span class="va">molb</span>, <span class="va">nmf2</span><span class="op">$</span><span class="va">W</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">newdf3</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">##          molb         X1         X2         X3         X4</span></span>
<span><span class="co">## 01005 BCR/ABL 0.13330363 0.07871361 0.08242624 0.12487073</span></span>
<span><span class="co">## 01010     NEG 0.07932313 0.10315914 0.07383619 0.07563357</span></span>
<span><span class="co">## 03002 BCR/ABL 0.13884623 0.08740193 0.05415050 0.15494375</span></span>
<span><span class="co">## 04007     NEG 0.11026347 0.10895646 0.03755877 0.11316966</span></span>
<span><span class="co">## 04008     NEG 0.09112892 0.07427573 0.07092190 0.04698532</span></span>
<span><span class="co">## 04010     NEG 0.08566357 0.08889185 0.07922343 0.05450893</span></span></code></pre>
<div class="sourceCode" id="cb36"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">nr3</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/randomForest/man/randomForest.html" class="external-link">randomForest</a></span><span class="op">(</span><span class="va">molb</span><span class="op">~</span><span class="va">.</span>, data<span class="op">=</span><span class="va">newdf3</span><span class="op">)</span></span>
<span><span class="va">nr3</span></span></code></pre></div>
<pre><code><span><span class="co">## </span></span>
<span><span class="co">## Call:</span></span>
<span><span class="co">##  randomForest(formula = molb ~ ., data = newdf3) </span></span>
<span><span class="co">##                Type of random forest: classification</span></span>
<span><span class="co">##                      Number of trees: 500</span></span>
<span><span class="co">## No. of variables tried at each split: 2</span></span>
<span><span class="co">## </span></span>
<span><span class="co">##         OOB estimate of  error rate: 17.12%</span></span>
<span><span class="co">## Confusion matrix:</span></span>
<span><span class="co">##         BCR/ABL NEG class.error</span></span>
<span><span class="co">## BCR/ABL      28   9   0.2432432</span></span>
<span><span class="co">## NEG          10  64   0.1351351</span></span></code></pre>
</div>
</div>
<div class="section level2">
<h2 id="upshots">Upshots<a class="anchor" aria-label="anchor" href="#upshots"></a>
</h2>
<p>This is a very informal overview of dimension reduction. The default choices for the Gajoux NMF implementation do seem inferior to the HALS implementation by Gregg. PCA seems useful for the task at hand.</p>
<p>Now we need to consider methylation data and issues of scale.</p>
</div>
  </main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside>
</div>



    <footer><div class="pkgdown-footer-left">
  <p>Developed by Vince Carey, Rpbert Gregg.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.1.</p>
</div>

    </footer>
</div>





  </body>
</html>
